### ISSUE TITLE
Cannot use legacy accessors on Blobs with > 4 axes

### ISSUE THREAD

**Issue by claudehang** [rohitgirdhar/GenerativePredictableVoxels#1]
Hi, when running the reconst.py, I encounter this error after configuring the necessary libs.

Can you give me any suggestion on how can I fix it? Many thanks!

I1028 17:29:32.755017  3432 net.cpp:413] Input 0 -> data
I1028 17:29:32.755089  3432 layer_factory.hpp:77] Creating layer conv1
I1028 17:29:32.755129  3432 net.cpp:106] Creating Layer conv1
I1028 17:29:32.755147  3432 net.cpp:454] conv1 <- data
I1028 17:29:32.755167  3432 net.cpp:411] conv1 -> conv1
I1028 17:29:32.757350  3432 net.cpp:150] Setting up conv1
I1028 17:29:32.757392  3432 net.cpp:157] Top shape: 1 96 55 55 (290400)
I1028 17:29:32.757408  3432 net.cpp:165] Memory required for data: 1161600
I1028 17:29:32.757436  3432 layer_factory.hpp:77] Creating layer relu1
I1028 17:29:32.757458  3432 net.cpp:106] Creating Layer relu1
I1028 17:29:32.757474  3432 net.cpp:454] relu1 <- conv1
I1028 17:29:32.757491  3432 net.cpp:397] relu1 -> conv1 (in-place)
I1028 17:29:32.757513  3432 net.cpp:150] Setting up relu1
I1028 17:29:32.757529  3432 net.cpp:157] Top shape: 1 96 55 55 (290400)
I1028 17:29:32.757544  3432 net.cpp:165] Memory required for data: 2323200
I1028 17:29:32.757557  3432 layer_factory.hpp:77] Creating layer pool1
I1028 17:29:32.757575  3432 net.cpp:106] Creating Layer pool1
I1028 17:29:32.757591  3432 net.cpp:454] pool1 <- conv1
I1028 17:29:32.757609  3432 net.cpp:411] pool1 -> pool1
I1028 17:29:32.757637  3432 net.cpp:150] Setting up pool1
I1028 17:29:32.757657  3432 net.cpp:157] Top shape: 1 96 27 27 (69984)
I1028 17:29:32.757669  3432 net.cpp:165] Memory required for data: 2603136
I1028 17:29:32.757683  3432 layer_factory.hpp:77] Creating layer norm1
I1028 17:29:32.757704  3432 net.cpp:106] Creating Layer norm1
I1028 17:29:32.757719  3432 net.cpp:454] norm1 <- pool1
I1028 17:29:32.757735  3432 net.cpp:411] norm1 -> norm1
I1028 17:29:32.757761  3432 net.cpp:150] Setting up norm1
I1028 17:29:32.757777  3432 net.cpp:157] Top shape: 1 96 27 27 (69984)
I1028 17:29:32.757791  3432 net.cpp:165] Memory required for data: 2883072
I1028 17:29:32.757803  3432 layer_factory.hpp:77] Creating layer conv2
I1028 17:29:32.757829  3432 net.cpp:106] Creating Layer conv2
I1028 17:29:32.757845  3432 net.cpp:454] conv2 <- norm1
I1028 17:29:32.757879  3432 net.cpp:411] conv2 -> conv2
I1028 17:29:32.774193  3432 net.cpp:150] Setting up conv2
I1028 17:29:32.774226  3432 net.cpp:157] Top shape: 1 256 27 27 (186624)
I1028 17:29:32.774241  3432 net.cpp:165] Memory required for data: 3629568
I1028 17:29:32.774263  3432 layer_factory.hpp:77] Creating layer relu2
I1028 17:29:32.774283  3432 net.cpp:106] Creating Layer relu2
I1028 17:29:32.774302  3432 net.cpp:454] relu2 <- conv2
I1028 17:29:32.774319  3432 net.cpp:397] relu2 -> conv2 (in-place)
I1028 17:29:32.774340  3432 net.cpp:150] Setting up relu2
I1028 17:29:32.774358  3432 net.cpp:157] Top shape: 1 256 27 27 (186624)
I1028 17:29:32.774374  3432 net.cpp:165] Memory required for data: 4376064
I1028 17:29:32.774389  3432 layer_factory.hpp:77] Creating layer pool2
I1028 17:29:32.774410  3432 net.cpp:106] Creating Layer pool2
I1028 17:29:32.774425  3432 net.cpp:454] pool2 <- conv2
I1028 17:29:32.774443  3432 net.cpp:411] pool2 -> pool2
I1028 17:29:32.774467  3432 net.cpp:150] Setting up pool2
I1028 17:29:32.774487  3432 net.cpp:157] Top shape: 1 256 13 13 (43264)
I1028 17:29:32.774500  3432 net.cpp:165] Memory required for data: 4549120
I1028 17:29:32.774515  3432 layer_factory.hpp:77] Creating layer norm2
I1028 17:29:32.774536  3432 net.cpp:106] Creating Layer norm2
I1028 17:29:32.774551  3432 net.cpp:454] norm2 <- pool2
I1028 17:29:32.774569  3432 net.cpp:411] norm2 -> norm2
I1028 17:29:32.774591  3432 net.cpp:150] Setting up norm2
I1028 17:29:32.774610  3432 net.cpp:157] Top shape: 1 256 13 13 (43264)
I1028 17:29:32.774624  3432 net.cpp:165] Memory required for data: 4722176
I1028 17:29:32.774639  3432 layer_factory.hpp:77] Creating layer conv3
I1028 17:29:32.774667  3432 net.cpp:106] Creating Layer conv3
I1028 17:29:32.774684  3432 net.cpp:454] conv3 <- norm2
I1028 17:29:32.774703  3432 net.cpp:411] conv3 -> conv3
I1028 17:29:32.816196  3432 net.cpp:150] Setting up conv3
I1028 17:29:32.816229  3432 net.cpp:157] Top shape: 1 384 13 13 (64896)
I1028 17:29:32.816244  3432 net.cpp:165] Memory required for data: 4981760
I1028 17:29:32.816267  3432 layer_factory.hpp:77] Creating layer relu3
I1028 17:29:32.816285  3432 net.cpp:106] Creating Layer relu3
I1028 17:29:32.816301  3432 net.cpp:454] relu3 <- conv3
I1028 17:29:32.816318  3432 net.cpp:397] relu3 -> conv3 (in-place)
I1028 17:29:32.816339  3432 net.cpp:150] Setting up relu3
I1028 17:29:32.816357  3432 net.cpp:157] Top shape: 1 384 13 13 (64896)
I1028 17:29:32.816372  3432 net.cpp:165] Memory required for data: 5241344
I1028 17:29:32.816387  3432 layer_factory.hpp:77] Creating layer conv4
I1028 17:29:32.816411  3432 net.cpp:106] Creating Layer conv4
I1028 17:29:32.816428  3432 net.cpp:454] conv4 <- conv3
I1028 17:29:32.816448  3432 net.cpp:411] conv4 -> conv4
I1028 17:29:32.847182  3432 net.cpp:150] Setting up conv4
I1028 17:29:32.847221  3432 net.cpp:157] Top shape: 1 384 13 13 (64896)
I1028 17:29:32.847236  3432 net.cpp:165] Memory required for data: 5500928
I1028 17:29:32.847255  3432 layer_factory.hpp:77] Creating layer relu4
I1028 17:29:32.847275  3432 net.cpp:106] Creating Layer relu4
I1028 17:29:32.847295  3432 net.cpp:454] relu4 <- conv4
I1028 17:29:32.847316  3432 net.cpp:397] relu4 -> conv4 (in-place)
I1028 17:29:32.847337  3432 net.cpp:150] Setting up relu4
I1028 17:29:32.847357  3432 net.cpp:157] Top shape: 1 384 13 13 (64896)
I1028 17:29:32.847371  3432 net.cpp:165] Memory required for data: 5760512
I1028 17:29:32.847388  3432 layer_factory.hpp:77] Creating layer conv5
I1028 17:29:32.847409  3432 net.cpp:106] Creating Layer conv5
I1028 17:29:32.847425  3432 net.cpp:454] conv5 <- conv4
I1028 17:29:32.847445  3432 net.cpp:411] conv5 -> conv5
I1028 17:29:32.868224  3432 net.cpp:150] Setting up conv5
I1028 17:29:32.868255  3432 net.cpp:157] Top shape: 1 256 13 13 (43264)
I1028 17:29:32.868270  3432 net.cpp:165] Memory required for data: 5933568
I1028 17:29:32.868299  3432 layer_factory.hpp:77] Creating layer relu5
I1028 17:29:32.868320  3432 net.cpp:106] Creating Layer relu5
I1028 17:29:32.868336  3432 net.cpp:454] relu5 <- conv5
I1028 17:29:32.868355  3432 net.cpp:397] relu5 -> conv5 (in-place)
I1028 17:29:32.868376  3432 net.cpp:150] Setting up relu5
I1028 17:29:32.868392  3432 net.cpp:157] Top shape: 1 256 13 13 (43264)
I1028 17:29:32.868407  3432 net.cpp:165] Memory required for data: 6106624
I1028 17:29:32.868422  3432 layer_factory.hpp:77] Creating layer pool5
I1028 17:29:32.868444  3432 net.cpp:106] Creating Layer pool5
I1028 17:29:32.868459  3432 net.cpp:454] pool5 <- conv5
I1028 17:29:32.868479  3432 net.cpp:411] pool5 -> pool5
I1028 17:29:32.868504  3432 net.cpp:150] Setting up pool5
I1028 17:29:32.868522  3432 net.cpp:157] Top shape: 1 256 6 6 (9216)
I1028 17:29:32.868540  3432 net.cpp:165] Memory required for data: 6143488
I1028 17:29:32.868554  3432 layer_factory.hpp:77] Creating layer fc6
I1028 17:29:32.868579  3432 net.cpp:106] Creating Layer fc6
I1028 17:29:32.868615  3432 net.cpp:454] fc6 <- pool5
I1028 17:29:32.868638  3432 net.cpp:411] fc6 -> fc6
I1028 17:29:34.558751  3432 net.cpp:150] Setting up fc6
I1028 17:29:34.558848  3432 net.cpp:157] Top shape: 1 4096 (4096)
I1028 17:29:34.558862  3432 net.cpp:165] Memory required for data: 6159872
I1028 17:29:34.558899  3432 layer_factory.hpp:77] Creating layer relu6
I1028 17:29:34.558953  3432 net.cpp:106] Creating Layer relu6
I1028 17:29:34.558979  3432 net.cpp:454] relu6 <- fc6
I1028 17:29:34.558998  3432 net.cpp:397] relu6 -> fc6 (in-place)
I1028 17:29:34.559021  3432 net.cpp:150] Setting up relu6
I1028 17:29:34.559036  3432 net.cpp:157] Top shape: 1 4096 (4096)
I1028 17:29:34.559047  3432 net.cpp:165] Memory required for data: 6176256
I1028 17:29:34.559062  3432 layer_factory.hpp:77] Creating layer drop6
I1028 17:29:34.559089  3432 net.cpp:106] Creating Layer drop6
I1028 17:29:34.559103  3432 net.cpp:454] drop6 <- fc6
I1028 17:29:34.559120  3432 net.cpp:397] drop6 -> fc6 (in-place)
I1028 17:29:34.559140  3432 net.cpp:150] Setting up drop6
I1028 17:29:34.559155  3432 net.cpp:157] Top shape: 1 4096 (4096)
I1028 17:29:34.559167  3432 net.cpp:165] Memory required for data: 6192640
I1028 17:29:34.559180  3432 layer_factory.hpp:77] Creating layer fc7
I1028 17:29:34.559200  3432 net.cpp:106] Creating Layer fc7
I1028 17:29:34.559213  3432 net.cpp:454] fc7 <- fc6
I1028 17:29:34.559237  3432 net.cpp:411] fc7 -> fc7
I1028 17:29:35.191236  3432 net.cpp:150] Setting up fc7
I1028 17:29:35.191305  3432 net.cpp:157] Top shape: 1 4096 (4096)
I1028 17:29:35.191318  3432 net.cpp:165] Memory required for data: 6209024
I1028 17:29:35.191344  3432 layer_factory.hpp:77] Creating layer relu7
I1028 17:29:35.191366  3432 net.cpp:106] Creating Layer relu7
I1028 17:29:35.191385  3432 net.cpp:454] relu7 <- fc7
I1028 17:29:35.191404  3432 net.cpp:397] relu7 -> fc7 (in-place)
I1028 17:29:35.191426  3432 net.cpp:150] Setting up relu7
I1028 17:29:35.191442  3432 net.cpp:157] Top shape: 1 4096 (4096)
I1028 17:29:35.191453  3432 net.cpp:165] Memory required for data: 6225408
I1028 17:29:35.191468  3432 layer_factory.hpp:77] Creating layer drop7
I1028 17:29:35.191485  3432 net.cpp:106] Creating Layer drop7
I1028 17:29:35.191498  3432 net.cpp:454] drop7 <- fc7
I1028 17:29:35.191515  3432 net.cpp:397] drop7 -> fc7 (in-place)
I1028 17:29:35.191534  3432 net.cpp:150] Setting up drop7
I1028 17:29:35.191548  3432 net.cpp:157] Top shape: 1 4096 (4096)
I1028 17:29:35.191560  3432 net.cpp:165] Memory required for data: 6241792
I1028 17:29:35.191573  3432 layer_factory.hpp:77] Creating layer fc8-imfeat
I1028 17:29:35.191593  3432 net.cpp:106] Creating Layer fc8-imfeat
I1028 17:29:35.191606  3432 net.cpp:454] fc8-imfeat <- fc7
I1028 17:29:35.191624  3432 net.cpp:411] fc8-imfeat -> fc8-imfeat
I1028 17:29:35.201620  3432 net.cpp:150] Setting up fc8-imfeat
I1028 17:29:35.201645  3432 net.cpp:157] Top shape: 1 64 (64)
I1028 17:29:35.201658  3432 net.cpp:165] Memory required for data: 6242048
I1028 17:29:35.201673  3432 layer_factory.hpp:77] Creating layer small-2
I1028 17:29:35.201689  3432 net.cpp:106] Creating Layer small-2
I1028 17:29:35.201704  3432 net.cpp:454] small-2 <- fc8-imfeat
I1028 17:29:35.201719  3432 net.cpp:411] small-2 -> small-2
I1028 17:29:35.202283  3432 net.cpp:150] Setting up small-2
I1028 17:29:35.202304  3432 net.cpp:157] Top shape: 1 216 (216)
I1028 17:29:35.202316  3432 net.cpp:165] Memory required for data: 6242912
I1028 17:29:35.202337  3432 layer_factory.hpp:77] Creating layer reshape
I1028 17:29:35.202359  3432 net.cpp:106] Creating Layer reshape
I1028 17:29:35.202373  3432 net.cpp:454] reshape <- small-2
I1028 17:29:35.202386  3432 net.cpp:411] reshape -> small-reshaped
I1028 17:29:35.202409  3432 net.cpp:150] Setting up reshape
I1028 17:29:35.202422  3432 net.cpp:157] Top shape: 1 1 6 6 6 (216)
I1028 17:29:35.202443  3432 net.cpp:165] Memory required for data: 6243776
I1028 17:29:35.202456  3432 layer_factory.hpp:77] Creating layer deconv4_3d
I1028 17:29:35.202478  3432 net.cpp:106] Creating Layer deconv4_3d
I1028 17:29:35.202491  3432 net.cpp:454] deconv4_3d <- small-reshaped
I1028 17:29:35.202508  3432 net.cpp:411] deconv4_3d -> deconv4_3d
I1028 17:29:35.202813  3432 net.cpp:150] Setting up deconv4_3d
I1028 17:29:35.202834  3432 net.cpp:157] Top shape: 1 256 8 8 8 (131072)
I1028 17:29:35.202847  3432 net.cpp:165] Memory required for data: 6768064
I1028 17:29:35.202862  3432 layer_factory.hpp:77] Creating layer derelu4_3d
I1028 17:29:35.202879  3432 net.cpp:106] Creating Layer derelu4_3d
I1028 17:29:35.202891  3432 net.cpp:454] derelu4_3d <- deconv4_3d
I1028 17:29:35.202905  3432 net.cpp:397] derelu4_3d -> deconv4_3d (in-place)
F1028 17:29:35.202927  3432 blob.hpp:140] Check failed: num_axes() <= 4 (5 vs. 4) Cannot use legacy accessors on Blobs with > 4 axes.
**\* Check failure stack trace: ***
Aborted (core dumped)

**Comment by rohitgirdhar** [comment]
This might have something to do with the caffe/cudnn version. Are you using the exact same version as mentioned in the readme? If not, you might want to google this error and follow the caffe issues, such as https://github.com/BVLC/caffe/pull/3983. Would be great if you can post your findings here!

**Comment by claudehang** [comment]
I've tried different version of caffe that seems to support 3D pooling, but still have this error. I know the error is with the 3D pooling, which is, specifically, the 3D PReLU layer. Could you please give me some suggestion on correctly using 3D PReLU? Very appreciate it for I am just novice caffe user.

**Comment by ghost** [comment]
Hi, I met the same error. I guess i may need to install a lower version cudnn libarary.
`Check failed: num_axes() <= 4 (5 vs. 4) Cannot use legacy accessors on Blobs with > 4 axes.
*** Check failure stack trace: ***
Aborted (core dumped)`

**Comment by ghost** [comment]
Hi, @rohitgirdhar  could you please details your development environment? thanks.

**Comment by rohitgirdhar** [comment]
I was using CUDA 7.0 with cuDNN 4

**Comment by ghost** [comment]
Oh, i have cuda 7.5 with cudnn 4, but it stills gives the same error. Could any one help me?

**Comment by YuBPan** [comment]
Hi, for the "check failed" issue, just comment line 140,141 in caffe/include/caffe/blob.hpp
```cpp
///    CHECK_LE(num_axes(), 4)
///        << "Cannot use legacy accessors on Blobs with > 4 axes.";
```
then recompile caffe again following the comment below.
Then it works.

**Comment by rohitgirdhar** [comment]
Thanks @YuBPan 
Closing this issue.

**Comment by haxis1** [comment]
@YuBPan sorry，the link in “this“ is broken. Can you renew it？ Thank you！

**Comment by YuBPan** [comment]
@haxis1 
1. fix (Caffe) check failed
  ```sh
  $ blob.hpp:140] Check failed: num_axes() <= 4 (5 vs. 4) Cannot use legacy accessors on Blobs with > 4 axes.
  *** Check failure stack trace: ***
  Aborted (core dumped)
  ```
  caffe/include/caffe/blob.hpp
  l140,141 comment them then recompile Caffe
  ```cpp
  ///    CHECK_LE(num_axes(), 4)
  ///        << "Cannot use legacy accessors on Blobs with > 4 axes.";
  ```
  [Caffe: ND convolution with im2col](https://github.com/BVLC/caffe/pull/2049)

2. show voxels outputs
  src/testing/reconst.py
  l52,53 add
  ```python
  import mayavi.mlab
  mayavi.mlab.points3d(out)
  ```

## results:

<img src="https://cloud.githubusercontent.com/assets/14142904/25065876/4f865c88-21e5-11e7-8357-ac17dcdb2af0.jpg" height="200px" /> --> <img src="https://cloud.githubusercontent.com/assets/14142904/25065873/47c09752-21e5-11e7-8b9a-0f6038fd02c7.png" height="200px" />

**Comment by haxis1** [comment]
@YuBPan thanks a lot!! I have already achieved the result. :-D

**Comment by thassanzadeh** [comment]
Hi, 
I commented line 140 and 141, but it shows the error again.
what is your mean recompile caffe??? I only changed the blob.hpp and saved it. but it shows error again. @haxis1 I installed caffe in anaconta by "coda install --name my_env caffe"

**Comment by rohitgirdhar** [comment]
You will need to compile caffe from source for this change to take effect. [Here](http://caffe.berkeleyvision.org/installation.html#compilation) are the instructions.

**Comment by thassanzadeh** [comment]
Hi, @rohitgirdhar 
Thank for your you reply.
as I said I installed caffe by :  coda install --name my_env caffe
at first I opened blob.hpp in the editor and commended lines 140 and 141 then
I went to my env and caffe rout and run following commands
(my_env) my@my-VirtualBox:~/caffe$ make all
(my_env) my@my-VirtualBox:~/caffe$ make test
(my_env) my@my-VirtualBox:~/caffe$ make runtest
but it shows the error again. what is the problem?
